### PR TITLE
Add initial support for extended names sections

### PR DESCRIPTION
**Pull request by sbc100**
See: https://github.com/WebAssembly/extended-name-section

Although this is only a phase 1 proposal its seems pretty
straight forward, and is already being implemented in
binaryen.

### CODE REVIEW

File: src/binary-reader.cc [hunk 1]
```diff
@@ -1731,6 +1736,32 @@ Result BinaryReader::ReadNameSection(Offset section_size) {
           }
         }
         break;
+      case NameSectionSubsection::Label:
+        // TODO(sbc): Implemnt label names. These are slightly more complicated
```
Comment (aardappel) [inline src/binary-reader.cc:1740]: `Implemnt`, `offers`

File: src/binary-writer.cc [hunk 2]
```diff
@@ -1514,6 +1516,155 @@ Result BinaryWriter::WriteModule() {
       }
     }
     EndSubsection();
+
+    size_t named_types = 0;
```
Comment (aardappel) [inline src/binary-writer.cc:1520]: the code blocks below all look very similar.. is it practical to pull this into a function?

**Review by aardappel — APPROVED** [review]

File: src/binary-writer.cc (hunk 2 above)
Comment (sbc100) [reply src/binary-writer.cc:1520]: Ahh yes... I guess I'll have to you one of those new-fangled template things I've heard about.

Comment (sbc100) [reply src/binary-writer.cc]: Done!

**Pull request merged**